cli/util/uploader.go

```diff
@@ -0,0 +1,429 @@
+				} else {
+					return res.startIndex, nil
+				}
+			}
+		}
+		if maxStartIndex != -1 {
+			return maxStartIndex + searchBatchSize, nil
+		}
+	}
+	return 0, nil
+}
+
+// updateIndexFiles updates the index files in the container.
+func updateIndexFiles(ctx *cli.Context, p *pool.Pool, containerID cid.ID, account wallet.Account, signer user.Signer, currentHeight uint, blockAttributeKey string, homomorphicHashingDisabled bool) error {
+	attributeKey := ctx.String("index-attribute")
+	indexFileSize := ctx.Uint("index-file-size")
+	fmt.Fprintln(ctx.App.Writer, "Updating index files...")
+
+	prm := client.PrmObjectSearch{}
+	filters := object.NewSearchFilters()
+	filters.AddFilter(attributeKey, fmt.Sprintf("%d", 1), object.MatchNumGE)
+	filters.AddFilter("size", fmt.Sprintf("%d", indexFileSize), object.MatchStringEqual)
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, p, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("search of index files failed: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / indexFileSize
+
+	if existingIndexCount >= expectedIndexCount {
+		fmt.Fprintf(ctx.App.Writer, "Index files are up to date. Existing: %d, expected: %d\n", existingIndexCount, expectedIndexCount)
+		return nil
+	}
+
+	var (
+		errCh                 = make(chan error)
+		buffer                = make([]byte, indexFileSize*oidSize)
+		oidCh                 = make(chan oid.ID, indexFileSize)
+		oidFetcherToProcessor = make(chan struct{}, indexFileSize)
+	)
+	defer close(oidCh)
+	for range maxParallelSearches {
+		go func() {
+			for currOid := range oidCh {
+				err := retry(func() error {
+					obj, err := p.ObjectHead(context.Background(), containerID, currOid, signer, client.PrmObjectHead{})
+					if err != nil {
+						return fmt.Errorf("failed to fetch object %s: %w", currOid.String(), err)
+					}
+
+					blockIndex, err := getBlockIndex(*obj, blockAttributeKey)
+					if err != nil {
+						return fmt.Errorf("failed to get block index from object %s: %w", currOid.String(), err)
+					}
+					offset := (uint(blockIndex) % indexFileSize) * oidSize
+					currOid.Encode(buffer[offset:])
+					oidFetcherToProcessor <- struct{}{}
+					return nil
+				})
+				if err != nil {
+					errCh <- err
+					return
+				}
+			}
+		}()
+	}
+
+	for i := existingIndexCount + 1; i <= expectedIndexCount; i++ {
+		startIndex := (i - 1) * indexFileSize
+		endIndex := startIndex + indexFileSize
+		go func() {
+			for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
+				remaining := int(endIndex) - j
+				end := j + min(searchBatchSize, remaining)
+
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
```

Comment:
Use `start <= x < end` pattern, it allows to avoid dances with indexes. Get rid of `remaining`. And this cycle won't work if `indexFileSize` is not a multiple of `searchBatchSize`.